src/goto-instrument/contracts/havoc_assigns_clause_targets.cpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************************\
+
+Module: Havoc multiple and possibly dependent targets simultaneously
+
+Author: Remi Delmas, delmasrd@amazon.com
+
+\*******************************************************************/
+
+/// \file
+/// Havoc multiple and possibly dependent targets simultaneously
+
+#include "havoc_assigns_clause_targets.h"
+
+#include <map>
+
+#include <langapi/language_util.h>
+
+#include <util/c_types.h>
+#include <util/format_expr.h>
+#include <util/format_type.h>
+#include <util/pointer_expr.h>
+#include <util/pointer_offset_size.h>
+#include <util/pointer_predicates.h>
+#include <util/std_code.h>
+
+#include "utils.h"
+
+static const char ASSIGNS_CLAUSE_REPLACEMENT_TRACKING[] =
+  " assigned by replaced ";
+
+/// Returns a pointer expression to the start address of the target
+///
+/// - If the target is a `pointer_object(p)` expression,
+///   return a pointer to the same object with offset zero
+/// - else, if the target is a sized and assignable expression,
+///   return its address
+/// - else trigger an error.
+///
+static exprt base_pointer(
+  const exprt &target,
+  // context parameters
+  const namespacet &ns)
+{
+  if(target.id() == ID_pointer_object)
+  {
+    const auto &ptr = target.operands().front();
+    // bring the offset back to zero
+    auto res = minus_exprt{ptr, pointer_offset(ptr)};
+    res.type() = ptr.type();
+    return std::move(res);
+  }
+  else if(is_assignable(target))
+  {
+    // check that the type of the target has a definite size
+    const auto &size = size_of_expr(target.type(), ns);
+
+    INVARIANT(
+      size.has_value(),
+      "`sizeof` must always be computable on assignable assigns clause "
+      "targets.");
+
+    return address_of_exprt{target, pointer_type(target.type())};
```

Comment:
Wouldn't `is_assignable` be true for something like `a[5]`, but taking the address thereof would not actually be the "base address" (or at least not how I'd understand that term).

Reply:
This computes the start address of the thing that gets assigned. If contract says it assigns `a[5]` then the base address of the thing that gets assigned is `&(a[5])`. If the contract says it assigns `__CPROVER_pointer_object(a[5])` then the base address will be the pointer equivalent to `a-5`.

Reply:
I'd then find something along the lines of "build_address_of" a more natural function name, but this might be a matter of personal preference.

Reply:
As a side, note that `address_of_exprt` has a convenience constructor that takes just the object, i.e., the above is equivalent to `address_of_exprt{target}`. The constructor will create the pointer type from the type of `target`.

Reply:
renamed to `build_address_of`